mne/io/edf/edf.py

```diff
@@ -215,22 +218,53 @@ def _read_segment(self, start=0, stop=None, sel=None, verbose=None,
                         step = block
                     samp = int(step * n_chan * data_size)
                     blocks = int(ceil(float(step) / sfreq))
-                    data = np.fromfile(fid, dtype=np.uint8, count=samp)
-                    data = data.reshape(-1, 3).astype(np.int32)
-                    # this converts to 24-bit little endian integer
-                    # # no support in numpy
-                    data = (data[:, 0] + (data[:, 1] << 8) +
-                            (data[:, 2] << 16))
-                    # 24th bit determines the sign
-                    data[data >= (1 << 23)] -= (1 << 24)
-
-                    data = data.reshape((int(sfreq), n_chan, blocks),
-                                        order='F')
                     for i in range(blocks):
-                        start_pt = int((sfreq * i) + (k * buffer_step))
-                        stop_pt = int(start_pt + sfreq)
-                        datas[:, start_pt:stop_pt] = data[:, :, i].T
+                        # complicated bdf: various sampling rates within file
+                        if 'n_samps' in self._edf_info:
+                            data = np.empty((n_chan, sfreq), dtype=int)
+                            for j, samp in enumerate(n_samps):
+                                chan_data = np.fromfile(fid, dtype=np.uint8,
+                                                        count=samp*data_size)
+                                chan_data = (chan_data[0::3] +
+                                             chan_data[1::3] << 8 +
+                                             chan_data[2::3] << 16)
+                                if j == stim_channel and samp < sfreq:
+                                    warnings.warn('Interpolating stim channel. '
+                                                  'Events may jitter.')
+                                    oldrange = np.linspace(0, 1, samp + 1,
+                                                           True)
+                                    newrange = np.linspace(0, 1, sfreq, False)
+                                    chan_data = interp1d(oldrange,
+                                                         np.append(chan_data, 0),
+                                                         kind='zero')(newrange)
+                                elif samp != sfreq:
+                                    mult = sfreq / samp
+                                    chan_data = resample(x=chan_data, up=mult,
+                                                         down=1, npad=0)
```

Comment:
for future compatibility (in case we decide to us polyphasic resampling instead of freq-domain) the `sfreq` and `samp` should be set to `up` and `down` (can't remember which is which) as opposed to setting `down=1` and using `up=ratio`.